### PR TITLE
fix #2007 http: Call callback in all cases

### LINKED ISSUE #2007: HTTP.get no callback on error
http.get on a local IP that is not available (not assigned) fails but does not call callback

### Expected behavior
The callback should be called  **allways**, even on all failures

in the example it should print
callback

### Actual behavior
The callback is not called

-- > http.get("http://192.1.1.254/",nil,function() print "callback" end)
-- > HTTP client: Disconnected with error: -11
-- HTTP client: Connection timeout
-- HTTP client: Connection timeout

### Test code

```Lua
-- http.get("http://192.1.1.254/",nil,function() print "callback" end)
```
### NodeMCU version
NodeMCU custom build by frightanic.com
	branch: master
	commit: c8ac5cfb912ff206b03dd7c60ffbb2dafb83fe5e
	SSL: true
	modules: bit,file,gpio,http,i2c,net,node,ow,pwm,rtctime,sjson,sntp,tmr,uart,wifi,tls
 build 	built on: 2017-06-17 17:53
 powered by Lua 5.1.4 on SDK 2.1.0(116b762)

### Hardware
Standard ESP6288 E-12 Dev Board

### PR DESCRIPTION
**Pull request by HHHartmann**
Call callback with errorcode -1 if no connection could be establioshed

Fixes #2007 and #1721.

Make sure all boxes are checked (add x inside the brackets) when you submit your contribution, remove this sentence before doing so.

- [x] This PR is for the `dev` branch rather than for `master`.
- [x] This PR is compliant with the [other contributing guidelines](https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md) as well (if not, please describe why).
- [x] I have thoroughly tested my contribution.
- [x] The code changes are reflected in the documentation at `docs/en/*`.

the PR checks the returnvalue of espconn_(secure_)disconnect. If no connection was established it fails and no disconnected callback is to be expected.
In this case the callback is called directly

I allso added some more error logging which I can remove on request.

### CODE REVIEW

File: app/http/httpclient.c [hunk 1]
```diff
@@ -451,20 +451,32 @@ static void ICACHE_FLASH_ATTR http_timeout_callback( void *arg )
...
+		HTTPCLIENT_ERR( "Connection is NULL" );
 		return;
 	}
 	if ( conn->reverse == NULL )
 	{
+		HTTPCLIENT_ERR( "reverse is NULL" );
 		return;
 	}
 	request_args_t * req = (request_args_t *) conn->reverse;
+	HTTPCLIENT_ERR( "Calling disconnect" );
```
Comment (marcelstoer) [inline app/http/httpclient.c]: Shouldn't this be `HTTPCLIENT_DEBUG`?

**Comment by marcelstoer** [comment]
Thanks you. In #2007 we established that there's also an inconsistency in our documentation wrt the disconnection handling. Can I ask you to clean this up as well in this PR?

File: app/http/httpclient.c [hunk 2]
```diff
@@ -451,20 +451,33 @@ static void ICACHE_FLASH_ATTR http_timeout_callback( void *arg )
 	struct espconn * conn = (struct espconn *) arg;
 	if ( conn == NULL )
 	{
+		HTTPCLIENT_DEBUG( "Connection is NULL" );
 		return;
 	}
 	if ( conn->reverse == NULL )
 	{
+		HTTPCLIENT_DEBUG( "reverse is NULL" );
```
Comment (marcelstoer) [inline app/http/httpclient.c]: It may be helpful to print a more explanatory message. The meaning of "reverse" may not be totally obvious.

Comment (HHHartmann) [inline app/http/httpclient.c]: @marcelstoer As to the documentation inconsistency: I am not aware of it, please enlighten me so I can add it.

The fix above establishes the behavior as in the docs AFAICS.

**Comment by HHHartmann** [comment]
@marcelstoer Documentation is a good point, already forgot about that problem. But there you go ...

**Comment by HHHartmann** [comment]
@marcelstoer well I intentionally left out the period as it had been omitted in the other paragraphs as well ;-)

**Pull request merged**